Governance changes: clearer timelines and conditions for decision making

fkiraly opened the pull request:
For consideragin in the 2021-07-05 CC meeting, I've drafted proposed changes to the governance guidelines that aim to make decision conditions for PR acceptance clearer and put clear timelines on decisions.

Major changes:
* making concrete conditions for core developers becoming active or inactive
* making explicit that core developer approval for proposals must be by a core developer who is not the proposer
* added lazy lazy approval for proposals made by core devs (new), so core dev PRs do not indefinitely lie around without approval - this is based on assumption that core devs know what they are doing
* clarified where approvals/rejections must be put
* replaced the gummi condition on "major change" (no definition, no decision maker who declares someting a "major change") with a clear rule on when sth is considered a "major change" - based on explicit core dev statement
* added timelines for decision making - 3 work days for non-major change, 12 work days for major change, 25 work days for stage 3 vote (if needed)
* clarifying what happens if lazy consensus is found (was not stated) - then PR must be merged by a core dev
* clarified that STEP must contain at least high-level design

fkiraly commented:
from CC meeting on 2021-07-05:
FK to clarify approval on PR vs positive vote
CC members to read and review until next meeting

mloning commented:
Hi all, here are a few thoughts on the suggested changes.

Regarding what happens if PRs do not receive any reply within a given amount of time: 

If a PR doesn't receive any replies, from my experience, it indicates that
* the proposed changes are perhaps not relevant at this point in time
* the changes have not been explained or communicated well
* there is something related to the PR makes people shy away from it

all of which for me would be indications to not merge the PR. 

Of course there may be a PR with none of the above problems that simply has not been commented on by anyone. In that case, to me the simpler solution would be to rely on core devs to communicate with each other and ask for reviews, rather than creating a rule where a PR gets merged automatically after certain amount of time.  

Regarding the right to have a review: I think we already cover this as a responsibility of the code owners, we can re-emphasise this, my main question about the right is how we would guarantee this and who would be responsible for guaranteeing this?

A few other suggestions: 
* rename "algorithm maintainers" to "code owners" (this is supported by GitHub and it's not just about algorithms)
* create speciality roles for different topics/modules who can help with triaging issues/PRs (thanks to @RNKuhns for the idea)

fkiraly commented:
>If a PR doesn't receive any replies, from my experience, it indicates that
> there is something related to the PR makes people shy away from it

Indeed, and that, I think is a key point that I want to address!

What I want to avoid is not developing a key area due to social dynamics unrelated to the content, like avoidance where there is no "perfect" solution (perfect is the enemy of good!), apathy/inertia, or proctrastination. 
A similar topic is disagreement between core devs where onlookers think providing a review means "picking a side". Complete avoidance of review responsibilities should not be "the easy way out" from difficult design or content discussions, dynamics like that can easily drive away active core contributors too (by getting used to that avoidance stance).

None of these dynamics are, in general, the pull requester's fault, they contribute to sclerotic decision making culture, and hence are typically to the detriment of health and development of sktime.

RNKuhns commented:
I personally think there are some things I think we could all do better as a group in terms of process and explanation. 

I'm willing to admit that given all the different developments with `sktime` that I will find myself unsure of what is being worked on, where I should focus my time (outside of functionality I'm working on adding) and most importantly how that fits into the larger plan. 

Many of us have a lot going on (e.g. full time jobs, completing degree programs, families, etc) and I think we should try and make it as easy as possible on everyone to focus their contribution time on substantive contribution (feedback/review, contributing issues/PRs, coordinating work, etc) as opposed to trying to stay caught up on everything that is going on. 

In terms of issues/PRs that are generated by the `sktime` development team, I'd like to see us have a more detailed roadmap of future development. This has a long-run component (in the long-run we'd like to have functionality X, Y, and Z) and a more tactical shorter-run component to help us coordinate our efforts better. This would include identifying areas of prioritized development in short-to-medium time horizons and having task boards for different things we'd like to add in different topic/module/functionality areas.

I think this helps because:

- It makes it much easier to understand what we plan to develop so developers can focus their efforts and therefore have more time for feedback/review, contributing issues/PRs, etc

- Authors of issues/PRs can link their explanations of the issues/PRs back to the plan for context and they can more easily focus on their explanation of what is most important for feedback and any important decision points they see related to the issue/PR


Note that I'm not advocating for us to rigidly just follow a set of task lists -- important and/or interesting things will pop-up along the way and will be worth working on in-lieu of what we planned on working on at that moment. The task lists and plans should be dynamic, we should just keep track of that dynamism.

Of course there will (hopefully) always be issues/PRs from the broader user community that will pop up. I think proposal on creating specialty roles for topics/modules (that was mentioned in @mloning's comment above) would help organize this process and reduce the burden on individual people (goal would be to try to have multiple people on topics/modules). The general idea on this front would be if an issue/PR comes in related to someone's topic/module area they'd engage on the PR provide initial triaging (is it actually a bug or is something else going on?, does the feature already exist or should it be added?, etc) and help shepherd along the issue/PR and make sure to get other core-developer input as needed.

fkiraly commented:
>I think proposal on creating specialty roles for topics/modules (that was mentioned in @mloning's comment above) would help organize this process and reduce the burden on individual people (goal would be to try to have multiple people on topics/modules). The general idea on this front would be if an issue/PR comes in related to someone's topic/module area they'd engage on the PR provide initial triaging (is it actually a bug or is something else going on?, does the feature already exist or should it be added?, etc) and help shepherd along the issue/PR and make sure to get other core-developer input as needed.

Hm, @RNKuhns, that's not a bad idea.

I would consider this is a separate matter though, i.e., introduction of a new role, and a proposal how this relates to PR reviews.
Do you maybe want to draft something and PR to the governance document?

Also, it's worth thinking about how much there is to be codified as "laws", and how much is better left to improvement of our project planning practices, e.g., via the community team.

Further, have you thought about what happens if new modules come up, or how this proposal of module or topic owners would interact with improvements to the infrastructure of the entire package? Think about the tag logic, for instance.

RNKuhns commented:
@fkiraly I'll work something up on the roles and talk with the community-team about it. I also agree that the first part is mainly about organization on all of our ends and that is also something the community-team can work on.

TonyBagnall commented:
I'll split reply into two, first bits I think are not contentious 
> * making concrete conditions for core developers becoming active or inactive
agreed
> * making explicit that core developer approval for proposals must be by a core developer who is not the proposer
agreed

> * clarified where approvals/rejections must be put
"Votes are cast publicly, in the location of the proposed change"
I'm not sure what "in the location of the proposed change" means. Votes are not anonymous? Could we not run a poll instead?

> There's one other unrelated change to the sentence at the start, about welcoming people "who have experienced discrimination" in the past. I'm not sure it's a good idea do define groups by their suffering rather than self-stated labels or qualifiers, so I changed it to "groups underrepresented in the tech sector".
agreed.

TonyBagnall commented:
the rest relates to PRs from core developers which are neither approved nor vetoed. On reflection, I am personally not in favour of the whole lazy approval mechanism as described. I would prefer rules along the lines of all PRs should receive a review within X days and if this does not happen the PR will automatically be put on the agenda for the next CC meeting or technical meeting where a review will be solicited

mloning commented:
Re the lazy lazy consensus, I like @TonyBagnall suggestion: if a PR by a core dev hasn't received any replies within X days, it will be put on the agenda in the next technical discussion meeting, if that doesn't happen it is escalated to the next CC meeting.

mloning commented:
My preferred solution: If a PR by a core developer has not received any replies within 10 working days, the author of the PR has the right to put the PR up for discussion as the top agenda item in the next technical committee.

fkiraly commented:
>My preferred solution: If a PR by a core developer has not received any replies within 10 working days, the author of the PR has the right to put the PR up for discussion as the top agenda item in the next technical committee.

Ok, that's what we agreed on - more specifically, it goes at the top of the agenda for *decision*.
Also, we don't have a technical committee yet, I assume it's the regular core dev meeting, right?

mloning commented on GOVERNANCE.rst:
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
+   two conditions must be satisfied:
+    - approval by at least one core developer and no rejection by a core developer (veto right).
+      The approval required for this condition must be by a core developer different from a proposer.
+    - approval in the regular core developer meeting as part of the agenda.
+      This can happen only if no approvals or rejections are expresesd on a PR, see below.
```
Let's rely on explicit approvals on GitHub
```suggestion
```
Also makes it easier: you need one approval (lazy consensus) and no rejection (veto right)

fkiraly replied:
but that's still an option! I'm just spelling out the approval conditions. Either what you say (approval, no veto), or approval in the core dev meeting if neither approval nor veto appears on the PR.

mloning commented on GOVERNANCE.rst:
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
+-  For a proposed change to be rejected via lazy consensus, one of the following two conditions
+   must be satisfied:
+    - rejection by at least one core developer, and no acceptance by a core developer.
+    - rejection in the regular core developer meeting as part of the agenda.
+      This can happen only if no approvals or rejections are expresesd on a PR, see below.
```
```suggestion
```

fkiraly replied:
no? I thought we agreed that if no approvals/rejections appear, it is tabled at the top of agenda at the core dev meeting.

fkiraly replied:
simplified

mloning commented on GOVERNANCE.rst:
```diff
@@ -384,24 +431,32 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
 comparison with alternative solutions, as outlined in our
 `template <https://github.com/sktime/enhancement-proposals/blob/master/TEMPLATE.md>`__.
 
+A complete STEP must always include at least a high-level design for the proposed change,
```
I think more info on what consistute a (good) STEP should go into the README of the enhancement-proposal repo.

fkiraly replied:
yes, but that's additional work, right? Should we open an issue?

mloning commented on GOVERNANCE.rst:
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
    detailed analysis and a consensus that is both explicit and informed.
-   These changes require an :ref:`steps`.
+   A major change always require a worked out :ref:`steps`.
+   A change is major in the above sense
+   if at least three core developers explicitly state that the proposal is a major change
+   and requires especial consideration, in the location.
```
I think some changes are major changes even if less than three core devs explicitly state that on the PR/issue. I'd say we leave out the definition of what counts as "major" here and resolve that through discussion if unclear. 
```suggestion
```

fkiraly replied:
well, but if we don't define what is "major", we cannot make "major" a condition in the rules, since it is undefined and arbitrary otherwise. It appears as a condition for PR that require a STEP.

So either we remove both the requirement of STEPS too, or we leave both things in.

fkiraly replied:
deferred to the table

mloning commented on GOVERNANCE.rst:
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
+   and requires especial consideration, in the location.
+-  Proposers are expected to give reasonable time for consideration, that is,
+   time and opportunity for core developers to review and
+   give their opinion on the contribution.
```
```suggestion
   give their opinion on the pull request.
```

fkiraly replied:
not all contributions are pull requests - or do we assume they are and must be?

fkiraly replied:
changed all contribs to PR

mloning commented on GOVERNANCE.rst:
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
+   Ten working days excluding week-ends constitute "reasonable time" in the above sense.
+   For pull requests, the period resets at every new change made to the PR.
+   It starts only when all GitHub checks pass.
+-  If the "reasonable time" period elapses and no approval or rejection has been expressed on a PR,
+   the PR is scheduled for decision, at the top of the next regular core developer meeting's agenda
+   in which less than seven PR are already on the agenda. If no regular core developer meeting
+   is scheduled within the two calendar weeks after the period elapses,
+   the core developers must schedule one within those two weeks.
```
I'd rephrase this as in my comment, since this only applies to PRs by core devs 
```suggestion
-  If a PR by a core developer has not received any approvals or rejections within ten working days excluding weekends, the core developer who opened the PR has the right to put the PR up for discussion at the top agenda in the next regular core developer meeting and receive a review from at least one of the other core developers. 
```

fkiraly replied:
this would make the process less clear:

* how do we prevent too long core dev meetings since there are so many PR?
* how do we prevent that no core dev meeting is planned, is it then indefinitely stalled?
* what does "review" mean? If it is a GitHub review, why is a core dev meeting needed? As opposed to a reviewer being appointed with the duty to review?

fkiraly replied:
clarified that the dev meeting appoints a reviewer.

mloning commented on GOVERNANCE.rst:
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
+   in which less than seven PR are already on the agenda. If no regular core developer meeting
+   is scheduled within the two calendar weeks after the period elapses,
+   the core developers must schedule one within those two weeks.
+-  Decision, in the core developer meeting, is by 2/3 majority of core developers in the meeting,
+   possibly including the proposer.
+   Decision is between approval or rejection. Approval can be subjected to minor corrections, which
+   requires an additional, simple majority consensus decision of core developers in the meeting.
+
+Approval or rejection as above count as "consensus".
+Lack of such consensus can arise under two conditions:
+- at least one approval and at least one rejection on GitHub
+- no 2/3 majority for either approval or rejection in the core developer meeting
```
```suggestion
```

mloning replied:
For me, the main purpose of adding it to the agenda is for the PR to receive at least some feedback, at this stage we don't have to vote yet, voting is also already handled in stage 2, so I'd remove these changes here.

fkiraly replied:
Not against this, but please note that by simply removing this, the process is incomplete. How does it reach stage 2 in your preferred process? There is no link between moving it to the core dev meeting and stage 2.

fkiraly replied:
simplified

mloning commented on GOVERNANCE.rst:
```diff
@@ -357,11 +403,12 @@ When no consensus can be found, any core developer can call for a vote
 -  All votes are a binary vote: for or against accepting the proposed
    changes.
 -  Votes are casts as comments: +1 (approval) or -1 (rejection).
+-  Votes are cast publicly, in the location of the proposed change.
```
```suggestion
-  Votes are cast publicly, on the pull request.
```

fkiraly replied:
are all proposals pull requests? See above.

mloning commented on GOVERNANCE.rst:
```diff
@@ -384,24 +431,32 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
 -  The CC will use consensus seeking.
--  If no consensus can be found within a month, the decision is made
-   through a simple majority vote (with tie breaking) among the CC
+-  If no consensus can be found within twenty working days excluding week-ends
```
```suggestion
-  If no consensus can be found within twenty working days excluding weekends
```

fkiraly replied:
appears I'm "old-fashioned", might have been reading too many old books.

A poster puts it nicely [here on quora](https://www.quora.com/Which-is-correct-weekend-or-week-end)

"
Weekend is in 2021 (and has been for many decades) the dominant form.

If you were to use week-end in writing, you would probably be labeled as old-fashioned or as fond of outdated language.

Still, there was a time when weekend was inconceivable, a time when it would have been almost obscene.

To write, around the 1940s:

Where have you been for the weekend?

Would've been just like saying, in 2021:

Who will be the garbage taker outer today?

(Which, by the way, you can translate into English by saying: Who will take out the garbage today?)

Probably one day such an expression will become standard English too.

Back to your question, the turning point was somewhere in the fifties, when weekend was introduced and gradually gained popularity over week-end.

Again, before this time week-end was the only form.
"

mloning reviewed (commented):
Hi @fkiraly, when you look at how the change render, the formatting looks a bit off, perhaps just a spacing issue, the rst format is sometimes a bit tricky.
 
A few comments below. Happy to help with these changes if you like.

fkiraly commented:
from discussion in 2021-08-09 CC meeting: Pr should not be for approval in core dev meeting, but for discussion & assigning review

fkiraly commented:
I've addressed the comments by simplifying the process:

* core dev meeting appoints a reviewer who needs to review and approve/reject
* there is only one condition for lazy consensus on approve or reject, so the process greatly simplifies
* "major change" definition is deferred to the table in the section before (which already was there, it is unchanged from pre-PR state); the concept of "major change" is removed, instead the table simply says when a STEP is needed

Due to the last point, I introduced a minor change:

* since all API changes would require STEPs strictly speaking by the table, I've made a modification which allows a STEP to be a short summary in an issue or PR if it is small; the CC decides when a longer STEP is needed.

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
 
 .. _stage-1:
 
-Stage 1: lazy consensus with veto right
```
I prefer the previous header since that's what it is

fkiraly replied:
ok

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
+   Some changes also require a worked out :ref:`steps`, this depends on the type of change, see
+   `decision making process <#Decision-making>`__ above.
+-  For a proposed change to be approved via lazy consensus, it needs to
+   approval by at least one core developer and no rejection by a core developer (veto right).
```
to explain the header 
```suggestion
   approval by at least one core developer (lazy consensus) and no rejection by a core developer (veto right).
```

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
-   detailed analysis and a consensus that is both explicit and informed.
-   These changes require an :ref:`steps`.
+-  Proposed changes should be in the form of GitHub pull requests (PR).
+   Some changes also require a worked out :ref:`steps`, this depends on the type of change, see
```
```suggestion
   Some changes also require a worked out :ref:`steps`. This depends on the type of change, see
```

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
+   `decision making process <#Decision-making>`__ above.
+-  For a proposed change to be approved via lazy consensus, it needs to
+   approval by at least one core developer and no rejection by a core developer (veto right).
+   The approval required for this condition must be by a core developer different from a proposer.
```
```suggestion
   Core developer cannot approve their proposed PRs. 
```

fkiraly replied:
this suggestion would not be clear in a case where multiple people work on a PR?

mloning replied:
if multiple core devs work on a PR and review each other code, that's fine no? of course it's even better if other core devs review too, especially if it's bigger API changes etc

fkiraly replied:
no, it's *not* fine in my opinion. You can't review your own stuff, it's not a review.

fkiraly replied:
let's discuss separately whether we allow PR authors to review their own PR, we can discuss at the next CC meeting. Until then, let's keep the status quo here.

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
+   The approval required for this condition must be by a core developer different from a proposer.
+-  For a proposed change to be rejected via lazy consensus, it needs to receive a
+   rejection by at least one core developer, and no acceptance by a core developer.
+-  Approvals must be in the form of a GitHub PR approval. of the PR in question.
```
```suggestion
-  Approvals must be in the form of a GitHub PR approval of the PR in question.
```

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
+-  For a proposed change to be rejected via lazy consensus, it needs to receive a
+   rejection by at least one core developer, and no acceptance by a core developer.
+-  Approvals must be in the form of a GitHub PR approval. of the PR in question.
+   Rejections can be expressed as -1 comments, or any written comments
+   containing "I formally reject" in the PR, in reference to it.
```
```suggestion
   Rejections must be in the form of a GitHub PR rejection on the PR in question.
```

fkiraly replied:
there is no such thing as a GitHub PR rejection?

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
+   Ten working days excluding week-ends constitute "reasonable time" in the above sense.
+   The period resets at every new change made to the PR.
+   It starts only when all GitHub checks pass.
+-  During this period, the PR can be merged if it has an approval and no rejection, but must be
+   reverted if it receives a rejection in addition.
```
I think this isn't practical and will introduce more confusion. 
```suggestion
-  During this period, the PR can be merged if it has an approval and no rejection, but must be
   reverted if it receives a rejection in addition.
```

fkiraly replied:
well, what's your alternative process? We must always wait for the period to elapse to merge? I don't think we have done this for a single PR.

The other alternative would leave a loophole where proposer and reviewer collude, the reviewer merging a few seconds after a PR appears; or, the less extreme but still problematic version of this where some core devs do not have enough time to review, e.g., since the PR happens on a weekend.

fkiraly replied:
The suggested change does not change anything, so I assume you are fine with the content.

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
+-  During this period, the PR can be merged if it has an approval and no rejection, but must be
+   reverted if it receives a rejection in addition.
+-  If the "reasonable time" period elapses and no approval or rejection has been expressed on a PR,
+   the PR is scheduled at the top of the next regular (bi-monthly) core developer meeting's agenda.
```
```suggestion
   the PR is scheduled at the top of agenda for the next developer meetup.
```
We should add a link to the meet up page on the new website here

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
+   reverted if it receives a rejection in addition.
+-  If the "reasonable time" period elapses and no approval or rejection has been expressed on a PR,
+   the PR is scheduled at the top of the next regular (bi-monthly) core developer meeting's agenda.
+   In that meeting, a core developer is assigned to review, and approve or reject within
+   5 days of the meeting, excluding week-ends.
```
```suggestion
   In that meeting, a core developer is assigned to review the PR and either approve or reject within five days of the meeting excluding weekends.
```

mloning commented on GOVERNANCE.rst:
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
+   In that meeting, a core developer is assigned to review, and approve or reject within
+   5 days of the meeting, excluding week-ends.
+
+Failure of lazy consensus, in the above sense, can arise only under the following condition:
+at least one approval and at least one rejection in the PR.
```
```suggestion
Failure of lazy consensus, in the above sense, can arise if a PR has received at least one approval and at least one rejection.
```

fkiraly replied:
it's not an "if" (sufficiency) but an "only if" (necessity), see here for terminology:
https://en.wikipedia.org/wiki/Necessary_and_sufficient_conditions#Definitions

(also, the "can" would be wrong in case you mean sufficiency)

mloning commented on GOVERNANCE.rst:
```diff
@@ -353,20 +380,23 @@ When no consensus can be found, the decision is escaled to :ref:`stage-2`.
-at any point during the discussion.
+Voting takes place:
+
+* when no lazy consensus can be found in stage 1 decision making
```
```suggestion
* when no lazy consensus can be found in stage 1 above
```

mloning commented on GOVERNANCE.rst:
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
 -  The CC will use consensus seeking.
--  If no consensus can be found within a month, the decision is made
-   through a simple majority vote (with tie breaking) among the CC
+-  If no consensus can be found within twenty working days excluding weekends
+   since the beginning of the stage-1 "reasonable time for consideration" period,
+   the decision is made through a simple majority vote (with tie breaking) among the CC
```
```suggestion
-  If no consensus can be found, the decision is made through a simple majority vote (with tie breaking) among the CC
```

mloning replied:
there may not always be a CC meeting within that period, the role of the CC meeting in that case would also be to try again to find a consensus, by perhaps taking parts of the PR and rejecting other, etc

fkiraly replied:
this would muddy the definition of what it means "no consensus can be found". When do we pass to the CC?

I really don't like that your suggested changes are reducing accountability and guarantee of speedy process. Whether you intend it or not, removal of clear condtions and clear decision making points is directly hampering effective governance.

mloning commented on GOVERNANCE.rst:
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
+-  If this vote is needed, the vote should take place within twenty-five working days after the beginning
+   of the stage-1 "reasonable time for consideration" period, during a regular, minuted CC meeting.
+-  Any proposal reaching stage 3 must be supported by an :ref:`steps`,
+   which has been made public at least 5 days, excluding week-ends, before the vote.
```
```suggestion
   which has been made public at least 5 days, excluding weekends, before the vote.
```

mloning commented on GOVERNANCE.rst:
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
+   the decision is made through a simple majority vote (with tie breaking) among the CC
    members.
--  Any CC decision must be supported by an :ref:`steps`, which has been made public and discussed before the vote.
+-  If this vote is needed, the vote should take place within twenty-five working days after the beginning
+   of the stage-1 "reasonable time for consideration" period, during a regular, minuted CC meeting.
```
```suggestion
```

fkiraly replied:
why would you want to remove this very generous time limit? It would introduce a problematic loophole where decision making can stall simply since the cc never votes on the PR.

We want to avoid dynamics where indefinite stalling of a decision is a viable option.

mloning replied:
I know, but it just adds to many conditions/rules/timescales which are hard to keep track of, the CC will be working to resolve conflicts, don't think we need to impose more time limits here

fkiraly replied:
ok, fine. Let's discuss this addition separately then, I still think giving timelines is important.

mloning commented on GOVERNANCE.rst:
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
 
-For all decision of the CC, an sktime enhancement proposal (STEP) must have been made
-public and discussed before the vote.
+sktime enhancmenet proposals (STEPs) are required for:
```
```suggestion
sktime enhancement proposals (STEPs) are required for:
```

mloning commented on GOVERNANCE.rst:
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
 `template <https://github.com/sktime/enhancement-proposals/blob/master/TEMPLATE.md>`__.
 
-We collect and discuss proposals in sktime’s `repository for
+A complete STEP must always include at least a high-level design for the proposed change,
```
As before, let's define STEPs in a single place in the README in the STEPS repo, so a link should do here.

fkiraly replied:
that would make that readme a formal document and an appendix to this one, are you sure?

mloning replied:
As said before, I don't see this as a legal document and I don't think it should be, so not really sure what "formal document" or "appendix" means, having a link should do the job

mloning replied:
I suggest we have the STEP guidelines on the website at: https://www.sktime.org/en/latest/enhancement_proposals.html

mloning replied:
So we can easily link it there

fkiraly replied:
yes, this is already linked, so I assume this is resolved.

mloning commented on GOVERNANCE.rst:
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
+Usually, we collect and discuss proposals in sktime’s `repository for
 enhancement-proposals <https://github.com/sktime/enhancement-proposals>`__.
 
+For smaller changes, such as punctual changes to the API or governance documents,
+the STEP can also be be part of an issue or pull request.
```
for smaller changes we don't need a STEP I think, every PR should have some description of what's going on, but that's not what we consider a STEP I'd say 
```suggestion
For smaller changes, such as punctual changes to the API or governance documents,
the STEP can also be be part of an issue or pull request.
```

fkiraly replied:
then this would be in contradiction with the pre-existing requirements for decision making, which require a STEP for every change to the API or the governance document.

mloning replied:
Yes every "bigger" change

fkiraly replied:
it doesn't say that in the table (section decision making)

mloning replied:
Sure, but I wouldn't call PR descriptions STEPS, so I'd say for smaller changes no STEP is needed

fkiraly replied:
ok - suggestion is identical to the original, so I assume this is resolved.

fkiraly commented:
I have now updated this with what I think was the remaining point outstanding (add the "veto right" explicitly back in the "lazy consensus" section header).

I've also moved the document to the new location, and updated the links where they were pointing to the roles list.
